core/shared/src/main/scala/org/typelevel/idna4s/core/CodePoint.scala

```diff
@@ -37,16 +37,25 @@ import cats.syntax.all._
  * text in ways we do not anticipate, such as bidi scripts or where the code point in isolation
  * is not able to be rendered at all.
  *
+ * The `toString` for this class does ''not'' render the `String` representation of this code
+ * point, it renders the base 10 and base 16 value of the code point, as well as the number of
+ * UTF-16 bytes (char values) needed to represent this code point and if possible a prose
+ * description of the code point. The reason for not rendering the `String` representation is
+ * that this type is often used in error cases where the code point will have no valid `String`
+ * representation, or where that representation is not meaningful, or where that representation
+ * could mangle the error message (as noted above).
+ *
+ * For example,
+ *
+ * {{{
+ * scala> CodePoint.fromChar('a')
+ * val res0: Either[String, org.typelevel.idna4s.core.CodePoint] = Right(CodePoint(value = 97, hexValue = 0x0061, name = LATIN SMALL LETTER A, charCount = 1))
+ * }}}
+ *
  * @see
  *   [[https://www.unicode.org/reports/tr36/ Unicode Security Considerations]]
  */
-sealed abstract class CodePoint extends Serializable {
-
-  /**
-   * The integral value of the code point.
-   */
-  def value: Int
-
+final case class CodePoint private (value: Int) extends AnyVal {
```

Comment:
Dumb question: does making this a `case class` leak a non-validating copy method? Can we just use an ordinary `class` here anyways? We are already overriding `toString`, and `equals` and `hashCode` are trivial.

Reply:
Good catch and yes we can do that.

Reply:
Actually I don't think we need to implement `equals` and `hashCode` anyway.
> User-defined value classes which avoid object allocation ... may not override equals or hashCode methods.

https://www.scala-lang.org/api/2.13.8/scala/AnyVal.html